File: backend/scala-seed-project/src/routes/RoleRoutes.scala
// src/routes/RoleRoutes.scala

import cats.effect._
import org.http4s._
import org.http4s.dsl.Http4sDsl
import org.http4s.HttpRoutes
import services.RoleService
import middleware.AuthMiddleware

class RoleRoutes(roleService: RoleService)(implicit F: Sync[IO]) extends Http4sDsl[IO] {
  private val authMiddleware = new AuthMiddleware()

  val routes: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ POST -> Root as user if authMiddleware.isAuthenticated(user) && authMiddleware.authorize(user, "createRole") =>
      for {
        roleData <- req.as[Role] // Assuming Role case class exists
        role <- roleService.createRole(roleData.name, roleData.permissions)
        response <- Ok(role)
      } yield response

    case DELETE -> Root / roleId as user if authMiddleware.isAuthenticated(user) && authMiddleware.authorize(user, "deleteRole") =>
      for {
        deletedCount <- roleService.deleteRoles(List(roleId.toInt))
        response <- Ok(s"Deleted $deletedCount roles.")
      } yield response

    case PATCH -> Root / roleId / "name" as user if authMiddleware.isAuthenticated(user) && authMiddleware.authorize(user, "changeRoleName") =>
      for {
        newName <- req.as[String] // Assuming new name is sent in the request body
        role <- roleService.changeRoleName(roleId.toInt, newName)
        response <- Ok(role)
      } yield response

    case PATCH -> Root / roleId / "permissions" as user if authMiddleware.isAuthenticated(user) && authMiddleware.authorize(user, "changeRolePermissions") =>
      for {
        newPermissions <- req.as[List[String]] // Assuming new permissions are sent in the request body
        role <- roleService.changeRolePermissions(roleId.toInt, newPermissions)
        response <- Ok(role)
      } yield response

    case GET -> Root as user if authMiddleware.isAuthenticated(user) && authMiddleware.authorize(user, "viewRoles") =>
      for {
        roles <- roleService.getAllRoles
        response <- Ok(roles)
      } yield response
  }
}

File: backend/node_version/src/routes/roleRoutes.ts
// routes/roleRoutes.ts
import { Router } from 'express';
import {
  createRoleHandler,
  deleteRolesHandler,
  changeRoleNameHandler,
  changeRolePermissionsHandler,
  getAllRolesHandler,
} from '../controllers/roleController';
import { isAuthenticated } from '../middleware/authenticateMiddleware';
import { authorizeMiddleware } from '../middleware/authorizeMiddleware';

const router = Router();

// Role management routes (admin-only)
router.post('/', isAuthenticated, authorizeMiddleware(['createRole']), createRoleHandler); // Create a new role
router.delete('/:roleId', isAuthenticated, authorizeMiddleware(['deleteRole']), deleteRolesHandler); // Delete a role
router.patch('/:roleId/name', isAuthenticated, authorizeMiddleware(['changeRoleName']), changeRoleNameHandler); // Change a role's name
router.patch('/:roleId/permissions', isAuthenticated, authorizeMiddleware(['changeRolePermissions']), changeRolePermissionsHandler); // Update a role's permissions
router.get('/', isAuthenticated, authorizeMiddleware(['viewRoles']), getAllRolesHandler); // Get all roles


export default router;
